src/dowel/csv_output.py

```diff
@@ -41,13 +42,25 @@ def record(self, data, prefix=''):
                     extrasaction='ignore')
                 self._writer.writeheader()
 
-            if to_csv.keys() != self._fieldnames:
-                self._warn('Inconsistent TabularInput keys detected. '
-                           'CsvOutput keys: {}. '
-                           'TabularInput keys: {}. '
-                           'Did you change key sets after your first '
-                           'logger.log(TabularInput)?'.format(
-                               set(self._fieldnames), set(to_csv.keys())))
+            # add new keys to _fieldnames set
+            old_header_size = len(self._fieldnames)
+            for key in to_csv.keys():
+                self._fieldnames.add(key)
+
+            # if new keys are added, rewrite the csv file
+            if len(self._fieldnames) > old_header_size:
+                self._writer = csv.DictWriter(
+                    self._log_file,
+                    fieldnames=sorted(list(self._fieldnames)),
+                    extrasaction='ignore')
+
+                self._log_file.seek(0)
+                self._writer.writeheader()
+
+                with open(self._file_name, 'r') as old_file:
```

Comment:
Does this work? It's quite surprising to me that re-opening the log file after we've overwritten the header still works.

If this does work, why does it work?

Reply:
This does pass the tests I added, so I can say it works for small header sizes. I did another test run to verify that it also works for csv with larger header size (500) with this code:

```
logger.add_output(dowel.CsvOutput('out.csv'))
logger.log('Starting up...')
for i in range(500):
    tabular.record('new_data'+str(i), i)
    logger.log(tabular)
    logger.dump_all()
logger.remove_all()
```

I checked csv DictReader's [implementation](https://svn.python.org/projects/python/trunk/Lib/csv.py). DictReader returns an OrderedDict for each row. My understanding is that even though the header is re-written, the keys in old header exist in the new header, and DictReader is able to find the associated keys of each row cell. Cells created by the new keys in new header will be auto-populated and returned as well. So the OrderedDict read by DictReader can be used by DictWriter, because it contains the same set of keys.

However, I do agree that this looks unclear and I don't know if it will break in a real-world scenario. A "safe" solution is to write to a new temporary file, rename the file, then point self._writer to this new file and delete the old file. In fact this was my first implementation, I switched to this code because it operates on the same file and avoids using additional memory.

I am not very familiar with python low level implementation, but these are what I can tell based on the source code. Please correct me if I am wrong;)